Update spec (#69)
foxglove/mcap: docs/specification/README.md

# MCAP File Format Specification

> Status: DRAFT

## Overview

MCAP is a modular container file format for recording timestamped pub/sub messages with arbitrary serialization formats.

MCAP files are designed to work well under various workloads, resource constraints, and durability requirements.

- [Structure](#file-structure)
  - [Header](#header)
  - [Footer](#footer)
  - [Data Section](#data-section)
  - [Summary Section](#summary-section)
- [Records](#records)
  - [Header](#header-op0x00)
  - [Channel Info](#channel-info-op0x01)
  - [Message](#message-op0x02)
  - [Chunk](#chunk-op0x03)
  - [Message Index](#message-index-op0x04)
  - [Chunk Index](#chunk-index-op0x05)
  - [Attachment](#attachment-op0x06)
  - [Attachment Index](#attachment-index-op0x07)
  - [Statistics](#statistics-op0x08)
  - [Metadata](#metadata-op0x09)
  - [Summary Offset](#summary-offset-op0x0A)
  - [Footer](#footer-op0x7f)
- [Serialization](#serialization)

## File Structure

An MCAP file structure follows:

    <MAGIC><Header><Data section>[<Summary section>][<Summary Offset section>]<Footer><MAGIC>

> Data, Summary, and Summary Offset sections are structured as a sequence of records:

Author: ```suggestion
The Data, Summary, and Summary Offset sections are structured as sequences of **records**:
```

- Channel Info
- Chunk Index
- Attachment Index
- Statistics
- Metadata

All records in the summary section MUST be grouped by opcode.

> Why? Summary Offset records reference blocks of records with the same opcode. For more efficient indexing of the Summary Section, the summary section records should be in sequences of the same opcode.

Channel Info records in the summary are duplicates of Channel Info records throughout the Data section.

### Summary Offset Section

The optional summary offset section contains [Summary Offset](#summary-offset-op0x0A) records.

The summary offset section must only contain Summary Offset records referencing records in the summary section.

## Records

MCAP files may contain a variety of records. Records are identified by a single-byte **opcode**. Record opcodes in the range 0x01-0x7F are reserved for future MCAP format usage. 0x80-0xFF are reserved for application extensions and user proposals.

All MCAP records are serialized as follows:

    <record type><record length><record>

Record type is a single byte opcode, and record length is a uint64 value.

> Records may be extended by adding new fields at the end of existing fields. Readers should ignore any unknown fields.

Author: ```suggestion
Records may be extended by adding new fields at the end of existing fields. Readers should ignore any unknown fields.

The Footer record will not be extended.
```

### Header (op=0x00)

| Bytes | Name | Type | Description |
| --- | --- | --- | --- |
| 4 + N | profile | String | The profile is used for indicating requirements for fields throughout the file (encoding, user_data, etc). If the value matches one of the [well-known profiles][profiles], the file should conform to the profile. This field may also be supplied empty or containing a framework that is not one of those recognized. When specifying a custom profile prefer the `x-` prefix to avoid conflict with future well-known profiles. |
| 4 + N | library | String | freeform string for writer to specify its name, version, or other information for use in debugging |
| 4 + N | metadata | Array<Tuple<string, string>> | Example keys: robot_id, git_sha, timezone, run_id. |

### Channel Info (op=0x01)

> A Channel Info record defines an encoded stream of messages on a topic.

Author: I'm gonna stick with the `topic` language as written (not in the suggestion).

The wiki page on pubsub says:

>In a topic-based system, messages are published to "topics" or named logical channels. Subscribers in a topic-based system will receive all messages published to the topics to which they subscribe. The publisher is responsible for defining the topics to which subscribers can subscribe.

Channel Info records are uniquely identified within a file by their channel ID. A Channel Info record must occur at least once in the file prior to any message referring to its channel ID.

| Bytes | Name | Type | Description |
| --- | --- | --- | --- |
| 2 | id | uint16 | A unique identifier for this channel within the file. Channel Ids must start at 0 |
| 4 + N | name | String | The channel name. Sometimes called a topic. |
| 4 + N | message_encoding | String | Encoding for messages on this channel. The value should be one of the [well-known message encodings](./well-known-encodings.md). Custom values should use `x-` prefix. |

> | 4 + N | schema_format | String | Format for the schema. The value should be one of the [well-known schema formats](./well-known-schema-formats.md). Custom values should use the `x-` prefix. |

Author: ```suggestion
| 4 + N | schema_encoding | String | Format for the schema. The value should be one of the [well-known schema formats](./well-known-schema-formats.md). Custom values should use the `x-` prefix. |
```

| 4 + N | schema | uint32 lengh prefixed Bytes | Schema should conform to the schema_encoding. |
| 4 + N | schema_name | String | An identifier for the schema. The schema name should conform to any schema_encoding requirements. |

> | 4 + N | user_data | Array<Tuple<string, string>> | Metadata about this channel |

Author: ```suggestion
| 4 + N | metadata | Array<Tuple<string, string>> | Metadata about this channel |
```

| 4 + N | metadata | Array<Tuple<string, string>> | Example keys: robot_id, git_sha, timezone, run_id. |

### Channel Info (op=0x01)

A Channel Info record defines an encoded stream of messages on a topic.

Channel Info records are uniquely identified within a file by their channel ID. A Channel Info record must occur at least once in the file prior to any message referring to its channel ID.

| Bytes | Name | Type | Description |
| --- | --- | --- | --- |
| 2 | id | uint16 | A unique identifier for this channel within the file. Channel Ids must start at 0 |
| 4 + N | name | String | The channel name. Sometimes called a topic. |
| 4 + N | message_encoding | String | Encoding for messages on this channel. The value should be one of the [well-known message encodings](./well-known-encodings.md). Custom values should use `x-` prefix. |
| 4 + N | schema_format | String | Format for the schema. The value should be one of the [well-known schema formats](./well-known-schema-formats.md). Custom values should use the `x-` prefix. |
| 4 + N | schema | uint32 lengh prefixed Bytes | Schema should conform to the schema_encoding. |

> | 4 + N | schema_version | String | A unique string to identify this schema across files. i.e, a hash of the schema. |

Author: ```suggestion
```

### Chunk (op=0x03)

A Chunk contains a batch of channel info and message records. The batch of records contained in a chunk may be compressed or uncompressed.

All messages in the chunk must reference channel infos recorded prior in the file (or the current chunk).

| Bytes | Name | Type | Description | Example |
| --- | --- | --- | --- | --- |
| 8 | start_time | Timestamp | Earliest message record_time in the chunk. |
| 8 | end_time | Timestamp | Latest message record_time in the chunk. |
| 8 | uncompressed_size | uint64 | Uncompressed size of the `records` field. |
| 4 | uncompressed_crc | uint32 | CRC32 checksum of uncompressed `records` field. A value of zero indicates that CRC validation should not be performed. |
| 4 + N | compression | String | compression algorithm | i.e. `lz4`, `zstd`, `""`. An empty string indicates no compression. Refer to [well-known compression formats][compression formats]. |
| N | records | Bytes | Repeating sequences of `<record type><record length><record>`. Compressed with the algorithm in the "compression" field. |

### Message Index (op=0x04)

A Message Index record maps timestamps to message records within a chunk.

> A sequence of Message Index records occurs immediately after each chunk. Exactly one Message Index record must exist in the sequence for every channel on which a message occurs inside the chunk.

Author: We need to somehow state that there is a 1:1 of Message Index records to unique channels occuring in the chunk.

> A sequence of Message Index records occurs immediately after each chunk. Exactly one Message Index record must exist in the sequence for every channel on which a message occurs inside the chunk.

Reviewer: but if you're writing an unindexed file, then there isn't. right?

### Attachment (op=0x06)

Attachment records contain artifacts such as text, core dumps, or calibration data.

Attachment records must not appear within a chunk.

| Bytes | Name | Type | Description |
| --- | --- | --- | --- |
| 4 + N | name | String | Name of the attachment, e.g "scene1.jpg". |
| 8 | created_at | Timestamp | Time at which the attachment was created. |
| 8 | record_time | Timestamp | Time at which the attachment was recorded. |
| 4 + N | content_type | String | MIME Type (e.g "text/plain"). |
| 8 + N | data | uint64 length-prefixed Bytes | Attachment data. |
| 4 | crc | uint32 | CRC32 checksum of preceding fields in the record. A value of zero indicates that CRC validation should not be performed. |

### Attachment Index (op=0x07)

> An Attachment Index record contains the location of an attachment in the file. An Attachment Index record exists for every Attachment record in the file.

Author: Yes. Every record except Header and Footer is optional. How better to say that?

| Bytes | Name | Type | Description |
| --- | --- | --- | --- |
| 8 | message_count | uint64 | Number of messages in the file across all topics. |
| 4 | channel_count | uint32 | Number of channels in the file across all topics. |
| 4 | attachment_count | uint32 | Number of attachments in the file. |
| 4 | chunk_count | uint32 | Number of chunks in the file. |
| 4 + N | channel_message_counts | Map<uint16, uint64> | Mapping from channel ID to total message count for the channel. An empty map indicates this statistic is not available. |

When using a Statistics record with channel_message_counts, the Summary Data section MUST contain a copy of all Channel Info records. The Channel Info records MUST occur prior to the statistics record.

> Why? The typical usecase for tools is to provide a listing of the types and quantities of messages stored in the file. Without an easy to access copy of the Channel Info records, tools would need to linearly scan the file for Channel Info records to display what types of messages exist in the file.

### Metadata (op=0x09)

A metadata record records arbitrary key/value user data.

| Bytes | Name | Type | Description |
| --- | --- | --- | --- |

> | 4 + N | metadata | Array<Tuple<string, string>> | Example keys: robot_id, git_sha, timezone, run_id. |

Reviewer: ```suggestion
| 4 + N | metadata | Array<Tuple<string, string>> | Example keys: `robot_id`, `git_sha`, `timezone`, `run_id`. |
```

### Summary Offset (op=0x0A)

> A Summary Offset record contains the location of a group of records with the same opcode within the summary section.

Reviewer: Is it allowed to have more than one summary offset record with the same group opcode? Does the order matter?

> A Summary Offset record contains the location of a group of records with the same opcode within the summary section.

Author: Order does not matter. If you split your groups of records then you would need more than one summary offset record. But doing this would degrade performance for little reason. We might want to add language preventing that. I think there is language about this in the Summary section.

| Bytes | Name | Type | Description |
| --- | --- | --- | --- |
| 8 | group_opcode | uint64 | The opcode of all records in the group. |
| 8 | group_start | uint64 | Byte offset from the start of the file of the first record in the group. |
| 8 | group_end | uint64 | Byte offset from the start of the file of the last byte record in the group. |

### Footer (op=0x7f)

> A Footer record contains end of file information. It must be the last record in the file.

Reviewer: ```suggestion
A Footer record contains end-of-file information. It must be the last record in the file. Readers using the index to read the file will begin with by reading the footer and trailing magic.
```

| Bytes | Name | Type | Description |
| --- | --- | --- | --- |
| 8 | summary_start | uint64 | Byte offset of the start of file to first record in the summary section. If there are no records in the summary section this should be 0. |
| 8 | summary_offset_start | uint64 | Byte offset from the start of the file to the first Summary Offset record. If there are no Summary Offset records this value should be 0. |

> | 4 | crc | uint32 | A CRC32 of all bytes from the start of the Summary section up to the crc32 bytes. |

Reviewer: ```suggestion
| 4 | summary_crc | uint32 | A CRC32 of all bytes from the start of the Summary section up through the end of the previous field in the footer record. |
```

## Serialization

### Fixed-width types

> Multi-byte integers are serialized using little endian byte order.

Reviewer: ```suggestion
Multi-byte integers (uint16, uint32, uint64) are serialized using [little-endian byte order](https://en.wikipedia.org/wiki/Endianness).
```

### String

> Strings are serialized using a uint32 byte length followed by the utf-8 bytes.

Reviewer: ```suggestion
Strings are serialized using a uint32 byte length followed by the string data, which should be valid [UTF-8](https://en.wikipedia.org/wiki/UTF-8).
```

    <byte length><utf-8 bytes>

### Bytes

> Bytes is sequence of bytes

Reviewer: ```suggestion
Bytes is sequence of bytes with no additional requirements.
```

> Why? The typical usecase for tools is to provide a listing of the types and quantities of messages stored in the file. Without an easy to access copy of the Channel Info records, tools would need to linearly scan the file for Channel Info records to display what types of messages exist in the file.

### Metadata (op=0x09)

A metadata record records arbitrary key/value user data.

| Bytes | Name | Type | Description |
| --- | --- | --- | --- |
| 4 + N | metadata | Array<Tuple<string, string>> | Example keys: robot_id, git_sha, timezone, run_id. |

### Summary Offset (op=0x0A)

A Summary Offset record contains the location of a group of records with the same opcode within the summary section.

| Bytes | Name | Type | Description |
| --- | --- | --- | --- |
| 8 | group_opcode | uint64 | The opcode of all records in the group. |
| 8 | group_start | uint64 | Byte offset from the start of the file of the first record in the group. |
| 8 | group_end | uint64 | Byte offset from the start of the file of the last byte record in the group. |

> | 8 | statistics_end | uint64 | Byte offset from the start of the file of the last byte in the last record of the group. |

Author: ```suggestion
```

### Footer (op=0x7f)

A Footer record contains end of file information. It must be the last record in the file.

| Bytes | Name | Type | Description |
| --- | --- | --- | --- |
| 8 | summary_start | uint64 | Byte offset of the start of file to first record in the summary section. If there are no records in the summary section this should be 0. |
| 8 | summary_offset_start | uint64 | Byte offset from the start of the file to the first Summary Offset record. If there are no Summary Offset records this value should be 0. |

> | 8 | crc | uint32 | A CRC32 of all bytes from the start of the Summary section up to the crc32 bytes. |

Author: ```suggestion
| 4 | crc | uint32 | A CRC32 of all bytes from the start of the Summary section up to the crc32 bytes. |
```
